wandb/sdk/wandb_artifacts.py

```diff
@@ -997,16 +997,7 @@ def store_file(
         preparer: "StepPrepare",
         progress_callback: Optional["progress.ProgressFn"] = None,
     ) -> bool:
-        # write-through cache
-        cache_path, hit, cache_open = self._cache.check_md5_obj_path(
-            util.B64MD5(entry.digest),  # TODO(spencerpearson): unsafe cast
-            entry.size if entry.size is not None else 0,
-        )
-        if not hit and entry.local_path is not None:
-            with cache_open() as f:
-                shutil.copyfile(entry.local_path, f.name)
-            entry.local_path = cache_path
-
+        # TODO: document the purpose of the return value.
```

Comment:
Is this supposed to inform retry logic? I don't get why we return False if we upload, False if we don't have anything to upload, but True if we aren't supplied with a destination to upload to.

Reply:
I think the boolean is used to inform the file pusher's dedupe stats. There are three cases for this:

- Server didn't supply an upload URL. That means it already has the file, so we can return `True` to indicate this file was deduplicated away.
- There is no `local_path`, which means this was a reference file. Dedupe statistics aren't really relevant here so we return `False` but even if we returned `True` I _don't_ think it would matter because references are ignored for dedupe stats IIRC.
- It's neither a reference nor is it a file we're already uploaded, so we return `False` to indicate we weren't able to dedupe this file.